#!/usr/bin/env python

PACKAGE = "dji_m100_trajectory"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

#       	Name			Type      	Level	Description				Default 	Min	Max

gen.add(	"traj_on",		bool_t,		0,	"Trajectory on flag", 			False)
gen.add(	"max_z_on",		bool_t,		0,	"Set z refrence on at max_z",		False)
gen.add(	"lidar_on",		bool_t,		0,	"Lidar on flag", 			False)
gen.add(	"adaptive_yaw_on",	bool_t,		0,	"Adaptive yaw on flag", 		False)
gen.add(	"reg_on",		bool_t,		0,	"Regression on flag",			False)
gen.add(	"climb",		bool_t,		0,	"Climb to max_z", 			False)
gen.add(	"land",			bool_t,		0,	"Land to ground", 			False)
gen.add(	"change_z",		bool_t,		0,	"Change in height flag", 		False)
gen.add(	"point_tracking_on",		bool_t,		0,	"Flag to switch on point tracking", 		False)
gen.add("pub_on_setpoint_position",	bool_t,		0,	"Publish flag on /mavros/setpoint_position/local", False)

trajectory_enum = gen.enum([ 	gen.const("hover_origin", int_t, 0, "Hover Trajectory (at origin)"),
				gen.const("hover",        int_t, 1, "Hover Trajectory"),
				gen.const("hover_lidar",  int_t, 2, "Hover with Lidar Trajectory"),
                       		gen.const("circle",       int_t, 3, "Circle Trajectory"),
                       		gen.const("figure_8",     int_t, 4, "Figure 8 Trajectory"),
				gen.const("square",	  int_t, 5, "Square Trajectory"),
				gen.const("setpoint",	  int_t, 6, "Setpoint Trajectory"),
				gen.const("GP",	  	  int_t, 7, "Global Planner Trajectory"),],
                     		"An enum to set flight trajectory")
gen.add(	"traj_type", 		int_t, 		0, 	"Trajectory type", 			0, 	0, 	8, edit_method=trajectory_enum)

gen.add( 	"pos_pub_delay",	  int_t, 	0, 	"Delay in position publisher",          40, 	0, 	200)

gen.add( 	"max_z",		double_t, 	0, 	"Max height [m]",  		   	2.0, 	0, 	9.0)
gen.add( 	"x_hover",		double_t, 	0, 	"Hover X position [m]",  		0, 	-70.0, 	10.0)
gen.add( 	"y_hover",		double_t, 	0, 	"Hover Y position [m]",  		0, 	-10.0, 	40.0)
gen.add( 	"z_hover",		double_t, 	0, 	"Hover Z position [m]",  		0, 	-5.0, 	80.0)
gen.add( 	"yaw_hover",		int_t, 		0, 	"Hover yaw angle [deg]",  		0, 	-180, 	180)
gen.add( 	"wall_dist",		double_t, 	0, 	"Distance from wall [m]",  		4.5, 	0.5, 	10.0)
gen.add( 	"x_sp_start",		double_t, 	0, 	"Start setpoint X position [m]",  	0, 	-10.0, 	10.0)
gen.add( 	"y_sp_start",		double_t, 	0, 	"Start setpoint Y position [m]",  	0, 	-10.0, 	10.0)
gen.add( 	"z_sp_start",		double_t, 	0, 	"Start setpoint Z position [m]",  	2.0, 	0, 	10.0)
gen.add( 	"length",		double_t, 	0, 	"Length of the wall to traverse [m]",	2, 	0, 	30.0)
gen.add( 	"height",		double_t, 	0, 	"Height of the wall to traverse [m]",	1, 	0, 	10.0)
gen.add( 	"camera_V_FOV",		double_t, 	0, 	"Vertical FOV of the camera [deg]",	43.2, 	0, 	100.0)
gen.add( 	"req_V_overlap_percent",int_t, 		0, "Vertical overlap required for images [%]",	50, 	0, 	100)

gen.add( 	"del_z",		double_t, 	0, 	"Change in height [m]",  	   	0.3, 	0, 	1.0)
gen.add( 	"des_radius",		double_t, 	0, 	"Radius of Circle [m]",  	   	1.0, 	0.1, 	1000.0)
gen.add( 	"des_velocity",		double_t, 	0, 	"Absolute Velocity of Circle [m/s]",  	0.5, 	0.1, 	1.5)

gen.add( 	"climb_rate",		double_t, 	0, 	"Climbing rate",  		   	0.4, 	0.1, 	0.6)
gen.add( 	"land_rate",		double_t, 	0, 	"Landing rate",  		   	0.4, 	0.1, 	0.6)
gen.add( 	 "v_d",	                double_t, 	0, 	"Drone Velocity",  		   	1.0, 	1.0, 	50.0)
gen.add( 	 "px",	                double_t, 	0, 	"Point-to-view x coordinate",  		-8.0,  	-10.0, 	10.0)
gen.add( 	 "py",	                double_t, 	0, 	"Point-to-view y coordinate",  		0, 	-10.0, 	10.0)
gen.add( 	 "pz",	                double_t, 	0, 	"Point-to-view z coordinate",  		2, 	-10.0, 	10.0)
gen.add( 	 "nx",	                double_t, 	0, 	"Point-to-view x normal",  		1.0,  	-10.0, 	10.0)
gen.add( 	 "ny",	                double_t, 	0, 	"Point-to-view y normal",  		0.0, 	-10.0, 	10.0)
gen.add( 	 "nz",	                double_t, 	0, 	"Point-to-view z normal",  		0.0, 	-10.0, 	10.0)

exit(gen.generate(PACKAGE, "m100_trajectory_v2", "set_trajectory_v2"))
